File: netDxf/Objects/SunReferences.cs
using System;
using System.Collections.Generic;
using netDxf.Header;
using netDxf.Tables;
using netDxf.Entities;

namespace netDxf.Objects
{
    internal static class SunReferences
    {
        internal static bool IsHost(DxfObject host) { return host is VPort || host is View || host is Viewport; }
        internal static DxfDatabaseObject Get(DxfObject host)
        { return host is VPort vport ? vport.Sun : host is View view ? view.Sun : (host as Viewport)?.Sun; }
        internal static bool IsPresent(DxfObject host)
        { return host is VPort vport ? vport.SunHandlePresent : host is View view ? view.SunHandlePresent : host is Viewport viewport && viewport.SunHandlePresent; }
        internal static void Set(DxfObject host, DxfDatabaseObject sun, bool present = true)
        {
            if (host is VPort vport) { vport.Sun = sun; vport.SunHandlePresent = present; }
            else if (host is View view) { view.Sun = sun; view.SunHandlePresent = present; }
            else if (host is Viewport viewport) { viewport.Sun = sun; viewport.SunHandlePresent = present; }
            else throw new ArgumentException("SUN requires a VIEW, VPORT or VIEWPORT owner.", nameof(host));
        }
        internal static void CheckProfile(DxfObject host, DxfVersion version)
        {
            if (!IsHost(host)) throw new ArgumentException("SUN requires a VIEW, VPORT or VIEWPORT owner.", nameof(host));
            if (version < (host is View ? DxfVersion.AutoCad2010 : DxfVersion.AutoCad2007))
                throw new NotSupportedException("SUN requires R2007 or later; named VIEW ownership requires R2010 or later.");
        }
        internal static void CheckClone(DxfObject host)
        { if (Get(host) != null) throw new NotSupportedException("Clone the SUN ownership subtree explicitly with DxfObjectDatabase.CloneSun into a registered destination host."); }
        internal static void Validate(DxfObject host, DxfObjectDatabase database, List<string> errors)
        {
            if (!IsHost(host) || !IsPresent(host)) return;
            try { CheckProfile(host, database.Document.DrawingVariables.AcadVer); }
            catch (NotSupportedException error) { errors.Add(error.Message); }
            DxfDatabaseObject sun = Get(host);
            if (sun != null && (!(sun is DxfSun || sun is DxfOpaqueObject && sun.CodeName == "SUN") || !database.IsRegistered(sun) || !ReferenceEquals(sun.Owner, host)))
                errors.Add("Invalid reciprocal SUN ownership: " + host.Handle);
        }
    }
}
namespace netDxf.Tables
{
    public partial class VPort
    {
        /// <summary>Gets the owned SUN object, or null. Use Objects.SetSun, CloneSun and EraseOwnedTree to change the attachment.</summary>
        public netDxf.Objects.DxfDatabaseObject Sun { get; internal set; }
        internal bool SunHandlePresent { get; set; }
    }
    public partial class View
    {
        /// <summary>Gets the owned SUN object, or null. Use Objects.SetSun, CloneSun and EraseOwnedTree to change the attachment.</summary>
        public netDxf.Objects.DxfDatabaseObject Sun { get; internal set; }
        internal bool SunHandlePresent { get; set; }
    }
}
namespace netDxf.Entities
{
    public partial class Viewport
    {
        /// <summary>Gets the owned SUN object, or null. Use Objects.SetSun, CloneSun and EraseOwnedTree to change the attachment.</summary>
        public netDxf.Objects.DxfDatabaseObject Sun { get; internal set; }
        internal bool SunHandlePresent { get; set; }
    }
}

File: javascript/netDxf/Objects/SunReferences.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { VPort } from '../Tables/VPort.js';
import { View } from '../Tables/View.js';
import { Viewport } from '../Entities/Viewport.js';
import { DxfSun } from './DxfSun.js';
import { DxfOpaqueObject } from './DxfOpaqueObject.js';
import { DxfVersion } from '../Header/DxfVersion.js';
import { ArgumentException, NotSupportedException } from '../../runtime/Errors.js';
/** Internal host adapter. It observes caller-supplied registration, not a fabricated database. */
export class SunReferences {
  static IsHost(host) { return host instanceof VPort || host instanceof View || host instanceof Viewport; }
  static Get(host) { return this.IsHost(host) ? host.Sun : null; }
  static IsPresent(host) { return this.IsHost(host) && host.SunHandlePresent; }
  static Set(host,sun,present=true) {
    if(!this.IsHost(host)) throw new ArgumentException('SUN requires a VIEW, VPORT or VIEWPORT owner.','host');
    host.Sun=sun;host.SunHandlePresent=present;
  }
  static CheckProfile(host,version) {
    if(!this.IsHost(host)) throw new ArgumentException('SUN requires a VIEW, VPORT or VIEWPORT owner.','host');
    if(version < (host instanceof View ? DxfVersion.AutoCad2010 : DxfVersion.AutoCad2007))
      throw new NotSupportedException('SUN requires R2007 or later; named VIEW ownership requires R2010 or later.');
  }
  static CheckClone(host) {
    if(this.Get(host) !== null) throw new NotSupportedException('Clone the SUN ownership subtree explicitly with DxfObjectDatabase.CloneSun into a registered destination host.');
  }
  static Validate(host,database,errors) {
    if(!this.IsHost(host) || !this.IsPresent(host)) return;
    try { this.CheckProfile(host,database.Document.DrawingVariables.AcadVer); }
    catch(error) { if(!(error instanceof NotSupportedException)) throw error;errors.Add(error.message); }
    const sun=this.Get(host);
    if(sun !== null && (!(sun instanceof DxfSun || sun instanceof DxfOpaqueObject && sun.CodeName === 'SUN') || !database.IsRegistered(sun) || sun.Owner !== host))
      errors.Add('Invalid reciprocal SUN ownership: ' + (host.Handle ?? ''));
  }
}
